README.md
# minesweeper-rs  [![Build Status](https://travis-ci.org/Vinatorul/minesweeper-rs.svg)](https://travis-ci.org/Vinatorul/minesweeper-rs) [![](http://meritbadge.herokuapp.com/minesweeper)](https://crates.io/crates/minesweeper) [![license](http://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/Vinatorul/minesweeper-rs/blob/master/LICENSE) [![Join the chat at https://gitter.im/Vinatorul/minesweeper-rs](https://img.shields.io/badge/GITTER-JOIN%20CHAT-brightgreen.svg)](https://gitter.im/Vinatorul/minesweeper-rs?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)


Simple minesweeper in Rust based on piston_window

## How to Play

To win the game - open all cells with numbers. 

Controls:
* `R` - restart
* `H` - field height setup
* `W` - field width setup
* `M` - mines quantity setup
* `Arrows` to choose cell (or change settings in `setup` mode)
* `Space` to open cell
* `Ctrl` to mark cell
* `Esc` to exit
* `LMB` click on a cell to open
* `RMB` click on a cell to mark

**NOTE**: marked cells disallow opennig them, until mark will be removed.

**NOTE**: pressing `Space` or `LMB` on opened cell will open it's neighbors (Thanks to [waynenilsen](https://github.com/waynenilsen) and [xilec](https://github.com/xilec)!)

## Screenshots

In game screenshot (green cells are marked)
![minesweeper ingame](ingame.png)
**NOTE**: you can see counter of marked cells at the right of the window

Defeat (red cells are missed mines)
![minesweeper defeat](defeat.png)
**NOTE**: with `*` marked mines which ended the game

## License
`minesweeper` is licensed under the MIT license. Please read the LICENSE file in this repository for more information.

## Compiling

Follow these instructions to compile `minesweeper-rs`.

 1. Clone the project `$ git clone https://github.com/Vinatorul/minesweeper-rs && cd minesweeper-rs`
 2. Build the project `$ cargo build --release` (**NOTE:** There is a large performance differnce when compiling without optimizations, so I recommend alwasy using `--release` to enable to them)
 3. Once complete, the binary will be located at `target/release/minesweeper`

## Options

Pass `-h` to arguments or look here: 

```
USAGE:
    minesweeper [FLAGS] [OPTIONS]

FLAGS:
    -h, --help       Prints help information
        --oldOGL     set OpenGL version to 2.1
    -V, --version    Prints version information

OPTIONS:
    -f, --field <field>       field size "width"x"height"
        --maxFPS <max_fps>    set max fps
    -m, --mines <mines>       max mines
    -s, --size <size>         window size "width"x"height"
```

# How to Contribute

Contributions are always welcome! Please use the following guidelines when contributing to `minesweeper`

1. Fork `minesweeper`
2. Clone your fork (`git clone https://github.com/$YOUR_USERNAME/minesweeper-rs && cd minesweeper-rs`)
3. Create new branch (`git checkout -b new-branch`)
4. Make your changes, and commit (`git commit -am "your message"`)
 * I use a [conventional](https://github.com/ajoslin/conventional-changelog/blob/a5505865ff3dd710cf757f50530e73ef0ca641da/conventions/angular.md) changelog format so I can update my changelog using [clog](https://github.com/thoughtram/clog)
 * In addition to the conventions defined above, I also use `imp`, `wip`, `gr`.
 * Format your commit subject line using the following format: `TYPE(COMPONENT): MESSAGE` where `TYPE` is one of the following:
    - `feat` - A new feature
    - `imp` - An improvement to an existing feature
    - `perf` - A performance improvement
    - `tests` - Changes to the testing framework or tests only
    - `fix` - A bug fix
    - `refactor` - Code functionality doesn't change, but underlying structure may
    - `style` - Stylistic changes only, no functionality changes
    - `wip` - A work in progress commit (Should typically be `git rebase`'ed away)
    - `chore` - Catch all or things that have to do with the build system, etc
    - `gr` - Graphics changes
 * The `COMPONENT` is optional, and may be a single file, directory, or logical component. Can be omitted if commit applies globally
5. Run the tests (`cargo test`)
6. `git rebase` into concise commits and remove `--fixup`s (`git rebase -i HEAD~NUM` where `NUM` is number of commits back)
7. Push your changes back to your fork (`git push origin $your-branch`)
8. Create a pull request! (You can also create the pull request first, and we'll merge when ready. This a good way to discuss proposed changes.)

CHANGELOG.md
<a name="v1.2.1"></a>
##  (2016-01-05)


#### Improvements

*   Reimplementation of field generation to avoid game over on first click after game beginning ([0723d12c](https://github.com/Vinatorul/minesweeper-rs/commit/0723d12c42c78c169723414ad72b5dd3575a0163))



<a name="v1.2"></a>
##  (2015-10-26)


#### Improvements

*   Display game result on GUI. ([e99e267f](https://github.com/Vinatorul/minesweeper-rs/commit/e99e267f33abdd071a405ba4337d5cce7ab37557))

*	Implement redrawing when it needs for fixing. ([9793bc38](https://github.com/Vinatorul/minesweeper-rs/commit/9793bc38041836098792bb55e210d40b6583e43a))

#### Features

*   added the timer ([eaecc403](https://github.com/Vinatorul/minesweeper-rs/commit/eaecc4035720e1c47c3a1591745a60525edb7395))



<a name="v1.1"></a>
##  (2015-09-07)


#### Features

*   added a mine counter ([d2906e9e](https://github.com/Vinatorul/minesweeper-rs/commit/d2906e9e39794fca338adc6536b8bac05276ec21))
*   colored numbers ([eaf61bbc](https://github.com/Vinatorul/minesweeper-rs/commit/eaf61bbce8a8950a9ca856f3f5c2dc2c8bac4c59))
*   colored numbers ([b527f805](https://github.com/Vinatorul/minesweeper-rs/commit/b527f8056564b610949c25aab109dfbfb5ef8af0))
*   show the user which mine ended the game ([b2168f01](https://github.com/Vinatorul/minesweeper-rs/commit/b2168f01d3591b275fb48adb5c16b541f4014541))
*   when you click on a revealed number, it clicks on unrevealed neighbors as a shortcut. ([67942f5c](https://github.com/Vinatorul/minesweeper-rs/commit/67942f5cf8f701f851c3a282139838bf4dbb219d))
* **Game:**  Add ability to set max fps ([decbf729](https://github.com/Vinatorul/minesweeper-rs/commit/decbf729cf757ee65091d4716797e97d88de48d1))

#### Bug Fixes

*   minor bugs fixes ([6d04d05e](https://github.com/Vinatorul/minesweeper-rs/commit/6d04d05e32e65e8030f59d049cc42bbb13ef92cb))
*   fixed problems with OpenGL version ([0df987b8](https://github.com/Vinatorul/minesweeper-rs/commit/0df987b856bbd7aeb92a68ef28c8ca02cfb49127))
*   true number of mines on board, not the case before ([0eaad50e](https://github.com/Vinatorul/minesweeper-rs/commit/0eaad50ef2536e07dae9d28df31499c95b35c946))
*   fixed problems with OpenGL version ([c370245c](https://github.com/Vinatorul/minesweeper-rs/commit/c370245cb4fdcdd88bfbd4c8e4edf41add08f308))
*   fixed glutin 48 error ([7b45d2b3](https://github.com/Vinatorul/minesweeper-rs/commit/7b45d2b3f56aadb50f4c3e9e6ecce8a02ae30ff6))
* **game:**  went out of bounds in some cases during neighbor check ([95b61f3b](https://github.com/Vinatorul/minesweeper-rs/commit/95b61f3b7d6989143cf968f31fea6fc079af3bd8))



<a name="v1.0"></a>
## v1.0 (2015-08-30)


#### Features

* **Game:**
  *  add config via command line arguments ([05736cc5](https://github.com/Vinatorul/minesweeper-rs/commit/05736cc5538e248ef300912c9506c8fe72d858f9))
  *  added ability to mark cells ([3cb91ea4](https://github.com/Vinatorul/minesweeper-rs/commit/3cb91ea45e18d21ed81da0c56812dc5053d88aba))
  *  Added grid ([489ba90a](https://github.com/Vinatorul/minesweeper-rs/commit/489ba90acfdb5afe7f08b3e4418aea46b76e5657))
* **UI:**
  *  add ability to change settings ([12c8ead9](https://github.com/Vinatorul/minesweeper-rs/commit/12c8ead97c0aab1a15da3d1e2615e3281cefcb4f))
  *  added ui drawing ([e8c81d20](https://github.com/Vinatorul/minesweeper-rs/commit/e8c81d20cf24dfe5781ce1d9771911553459dacf))

#### Improvements

* **Field:**  allows field to fly ([730fac12](https://github.com/Vinatorul/minesweeper-rs/commit/730fac121f1dfbce530b6c835909dbd1f2b52028))
* **Game:**  added ability to win ([568b3a9c](https://github.com/Vinatorul/minesweeper-rs/commit/568b3a9c5ad4f43574ecf4d6c0e13d2745ade22c))



<a name="v0.1.0"></a>
## v0.1.0 (2015-08-27)


#### Improvements

* **Field:**  now clicking on empty cell will reveal all nearby cells ([d5dae61f](https://github.com/Vinatorul/minesweeper-rs/commit/d5dae61fc8922e1a9a0f933f56daba891ef6cb1f))
* **Game, Field:**  massive code improvements ([f4359660](https://github.com/Vinatorul/minesweeper-rs/commit/f4359660659dd88b3e755e2a2dbc6829f5551a95))

src/ui.rs

use piston_window::*;
use common::{ParamType, MoveDestination, GameEndState};
use chrono::Duration;

struct Block<'a> {
    name: &'a str, 
    num: u32,
    old_num: u32,
    hotkey: char,
    selected: bool,
    min_val: u32,
    max_val: u32,
    param_type: ParamType,
}

impl<'a> Block<'a> {
    pub fn new(name: &'a str,
               num: u32,
               hotkey: char,
               min_val: u32,
               max_val: u32,
               param_type: ParamType) -> Block<'a>
    {
        Block {
            name: name,
            num: num,
            old_num: num,
            hotkey: hotkey,
            selected: false,
            min_val: min_val,
            max_val: max_val,
            param_type: param_type,
        }
    }

    pub fn draw(&self,
                context: Context,
                graphics: &mut G2d,
                rect: &mut [u32; 4],
                glyps: &mut Glyphs)
    {
        let margin = 10;
        let text_height = 20;
        let block_height = 30;
        let text_padding = 7;
        let color = if self.selected {
            [0.0, 1.0, 0.0, 1.0]
        } else {
            [1.0, 1.0, 1.0, 1.0]
        };
        rect[1] += 20;
        rectangle(color,
                  [
                    (rect[0] + margin) as f64,
                    (rect[1]) as f64,
                    (rect[2] - 2*margin) as f64,
                    block_height as f64
                  ],
                  context.transform,
                  graphics);
        rect[1] += block_height;
        let transform = context.transform.trans((rect[0] + margin + text_padding) as f64,
                                                (rect[1] - text_padding) as f64);
        text([0.0, 0.0, 0.0, 1.0],
             text_height,
             &*format!("{}: {}", self.name, self.num),
             glyps,
             transform,
             graphics);
        rect[1] += margin + text_height/2;
        let transform = context.transform.trans((rect[0] + margin) as f64,
                                                (rect[1]) as f64);
        let s = if self.selected {
            format!("Use arrows, press '{}' to apply changes", self.hotkey)
        } else {
            format!("Press '{}' to change", self.hotkey)
        };
        text([1.0, 1.0, 1.0, 1.0],
             text_height,
             &*s,
             glyps,
             transform,
             graphics);
    }

    pub fn inc_safe(&mut self) {
        if self.num < self.max_val {
            self.num += 1;
        }
    }

    pub fn dec_safe(&mut self) {
        if self.num > self.min_val {
            self.num -= 1;
        }
    }

    pub fn apply_changes(&mut self) {
        self.old_num = self.num;
    }

    pub fn discard_changes(&mut self) {
        self.num = self.old_num;
    }
}

pub struct UI<'a> {
    blocks: Vec<Block<'a>>,
    selected_block: i32,
}    

impl<'a> UI<'a> {
    pub fn new(height: u32, width: u32, mines: u32) -> UI<'a> {
        UI {
            blocks: vec![Block::new("Field width", width, 'W', 5, 50, ParamType::Width),
                         Block::new("Field height", height, 'H', 5, 50, ParamType::Height),
                         Block::new("Mines", mines, 'M', 1, 2500, ParamType::Mines),
            ],
            selected_block: -1,
        }
    }

    pub fn draw(&mut self,
            context: Context,
            graphics: &mut G2d,
            mut rect: [u32; 4],
            glyps: &mut Glyphs,
            mines_total: u32,
            mines_marked: u32,
            duration : Duration)
    {
        for b in self.blocks.iter() {
            b.draw(context, graphics, &mut rect, glyps);
        }

        let transform = context.transform.trans((rect[0]+10) as f64,
                                                 (rect[1]+27) as f64);
        text([1.0, 1.0, 1.0, 1.0],
             20,
             &*format!("{} marked {} remaining", mines_marked, mines_total as i32 - mines_marked as i32),
             glyps,
             transform,
             graphics);
        
        let transform = context.transform.trans((rect[0]+10) as f64,
                                                 (rect[1]+27*2) as f64);
        let total_seconds = duration.num_seconds();
        let mins = total_seconds / 60;
        let rem_seconds = total_seconds - mins*60;

        text([1.0, 1.0, 1.0, 1.0],
             20,
             &*format!("{}:{}", mins, rem_seconds),
             glyps,
             transform,
             graphics);
    }

    pub fn proc_key(&mut self, block: ParamType) -> Option<u32> {
        if self.selected_block >= 0 {
            let selected = self.blocks.get_mut(self.selected_block as usize).unwrap();
            if selected.param_type == block {
                selected.selected = false;
                selected.apply_changes();
                self.selected_block = -1;
                return Some(selected.num);
            }
        }
        for i in 0..self.blocks.len() {
            let item = self.blocks.get_mut(i).unwrap();
            let b = item.param_type == block;
            item.selected = b;
            if b {
                self.selected_block = i as i32;
            } else {
                item.discard_changes();
            }
        }
        None
    }

    pub fn change_selected(&mut self, dest: MoveDestination) {
        assert!(self.selected_block >= 0);
        let selected = self.blocks.get_mut(self.selected_block as usize).unwrap();
        match dest {
            MoveDestination::Up | MoveDestination::Right => selected.inc_safe(),
            MoveDestination::Down | MoveDestination::Left => selected.dec_safe()
        }
    }
}

pub struct EndMessage<'a> {
    visible: bool,
    win: bool,
    message: &'a str
}

static EM_TEXT_BIG      : u32 = 40;
static EM_TEXT_SMALL    : u32 = 20;
static EM_TEXT_PADDING  : u32 = 5;
static EM_BORDER_WIDTH  : u32 = 3;
static EM_BORDER_WIDTH_2: u32 = 6;

static RETRY_MSG: &'static str = "Press 'R' to play again";

impl<'a> EndMessage<'a> {
    pub fn new() -> EndMessage<'a> {
        EndMessage {
            visible: false,
            win: false,
            message: "Ready",
        }
    }

    pub fn show( &mut self, end_state: GameEndState ) {
        self.win = match end_state {
            GameEndState::Win => true,
            _                 => false };
        self.visible = true;
        self.message = if self.win {
                "You WIN :)"
            } else {
                "Game over :("
            };
    }

    pub fn hide( &mut self ) {
        self.visible = false;
    }

    /// Message Box's size. (width,height).
    pub fn size() -> (u32,u32) {
        (
            350,
            (EM_TEXT_BIG + EM_TEXT_SMALL + (EM_TEXT_PADDING*4) + (EM_BORDER_WIDTH*2))
        )
    }

    pub fn draw(&self,
        context: Context,
        graphics: &mut G2d,
        rect: [u32; 4],
        glyps: &mut Glyphs)
    {
        if !self.visible
        {
            return;
        }

        let border_color = if self.win {
            [0.0, 0.0, 1.0, 1.0]    // Win : Blue
        } else {
            [1.0, 1.0, 0.0, 1.0]    // Lose: yellow
        };

        // draw border
        rectangle(border_color,
            [
                (rect[0]) as f64,
                (rect[1]) as f64,
                (rect[2]) as f64,
                (rect[3]) as f64
            ],
            context.transform,
            graphics);

        // draw background
        rectangle( [1.0, 1.0, 1.0, 1.0],
            [
                (rect[0] +EM_BORDER_WIDTH) as f64,
                (rect[1] +EM_BORDER_WIDTH) as f64,
                (rect[2] -EM_BORDER_WIDTH_2) as f64,
                (rect[3] -EM_BORDER_WIDTH_2) as f64
            ],
            context.transform,
            graphics);

        // Draw game result message
        let trans_msg = context.transform.trans(
            (rect[0] + EM_BORDER_WIDTH + EM_TEXT_PADDING) as f64,
            (rect[1] + EM_BORDER_WIDTH + EM_TEXT_PADDING + EM_TEXT_BIG) as f64);

        text([0.0, 0.0, 0.0, 1.0],
             EM_TEXT_BIG,
             &*self.message,
             glyps,
             trans_msg,
             graphics);

        // Draw 'play again' message
        let trans_retry = context.transform.trans(
            (rect[0] + EM_BORDER_WIDTH + EM_TEXT_PADDING) as f64,
            (rect[1] + EM_BORDER_WIDTH + EM_TEXT_PADDING*2 + EM_TEXT_BIG + EM_TEXT_SMALL) as f64);
        text([0.0, 0.0, 0.0, 1.0],
             EM_TEXT_SMALL,
             &*RETRY_MSG,
             glyps,
             trans_retry,
             graphics);
    }
}


src/game.rs
use field::{Field, Content};
use ui::UI;
use ui::EndMessage;
use piston_window::*;
use common::{ParamType, MoveDestination, GameEndState};
use chrono::{DateTime, UTC, Duration};  

enum RenderState {
    // redraw picture on current buffer
    First,

    // redraw picture on next buffer
    Second,

    // needn't redraw picture
    None
}

impl RenderState {

    // Update render state and return flag necessity of redrawing pickture
    fn update_state(&mut self) -> bool {
        match *self {
            RenderState::First  => *self = RenderState::Second,
            RenderState::Second => *self = RenderState::None,
            RenderState::None   => return false
        }

        true
    }

    fn reset(&mut self) {
        *self = RenderState::First
    }
}

pub struct Game<'a> {
    field: Field,
    ui: UI<'a>,
    msg: EndMessage<'a>,
    glyphs: Glyphs,
    mouse_x: f64,
    mouse_y: f64,
    game_ended: bool,
    panel_width: u32,
    in_ui: bool,
    game_start : Option<DateTime<UTC>>,
    last_time : Option<DateTime<UTC>>,
    game_end : Option<DateTime<UTC>>,
    // state of rendering for drawing picture on both buffers
    render_state: RenderState
}

impl<'a> Game<'a> {
    pub fn new(glyphs: Glyphs, width: u32, height: u32, mines: u32) -> Game<'a> {
        Game {
            field: Field::new(width, height, mines),
            ui: UI::new(width, height, mines),
            msg: EndMessage::new(),
            glyphs: glyphs,
            mouse_x: 0.0,
            mouse_y: 0.0,
            game_ended: false,
            panel_width: 350,
            in_ui: false,
            game_start: None,
            game_end: None,
            last_time: None,
            render_state: RenderState::First
        }
    }

    pub fn render(&mut self, window: &PistonWindow) {      
        if !self.render_state.update_state() {
            return;
        }

        window.draw_2d(|c, g| {
            clear([0.0, 0.0, 0.0, 1.0], g);
            let field_rect = self.get_field_rect(window);
            self.field.draw(c, g, field_rect, &mut self.glyphs);

            let ui_rect = self.get_ui_rect(window);

            let dur = match self.game_start {
                Some(game_start) => match self.game_end {
                    Some(game_end) => game_end - game_start,
                    None => UTC::now() - game_start,
                },
                None => Duration::seconds(0), 
            };

            self.ui.draw(c, g, ui_rect, &mut self.glyphs, self.field.total_mines(), self.field.count_marked(), dur);

            let msg_rect = self.get_msg_rect(window);
            self.msg.draw(c, g, msg_rect, &mut self.glyphs);
        });
    }

    fn get_field_rect(&self, window: &PistonWindow) -> [u32; 4] {
        let mut w = window.size().width - self.panel_width;
        w = (w /self.field.get_width()) * self.field.get_width();
        let mut h = window.size().height;
        h = (h /self.field.get_height()) * self.field.get_height();
        [0, 0, w, h]
    }

    fn get_ui_rect(&self, window: &PistonWindow) -> [u32; 4] {
        let mut field_w = window.size().width - self.panel_width;
        field_w = (field_w /self.field.get_width()) * self.field.get_width();
        let w = window.size().width - field_w;
        let h = window.size().height;
        [field_w, 0, w, h]
    }
    
    fn get_msg_rect(&self, window: &PistonWindow) -> [u32; 4] {
        let w_w = window.size().width;
        let w_h = window.size().height;
        let (m_w, m_h) = EndMessage::size();

        [
            ( (w_w-m_w) / 2 ),
            ( (w_h-m_h) / 2 ),
            m_w,
            m_h        
        ]
    }

    pub fn proc_key(&mut self, button: Button, window: &PistonWindow) {
        let mut need_redraw = true;
        if self.in_ui {
            match button {
                Button::Keyboard(key) => {
                    match  key {
                        Key::H => {
                            match self.ui.proc_key(ParamType::Height) {
                                Some(h) => {
                                    self.in_ui = false;                                    
                                    self.field.reinit_field(h, ParamType::Height);
                                    self.restart();
                                }
                                _ => need_redraw = false
                            }
                        },
                        Key::M => {
                            match self.ui.proc_key(ParamType::Mines) {
                                Some(m) => {
                                    self.in_ui = false;
                                    self.field.reinit_field(m, ParamType::Mines);
                                    self.restart();
                                }
                                _ => need_redraw = false
                            }
                        },
                        Key::W => {
                            match self.ui.proc_key(ParamType::Width) {
                                Some(w) => {
                                    self.in_ui = false;
                                    self.field.reinit_field(w, ParamType::Width);
                                    self.restart();
                                }
                                _ => need_redraw = false
                            }
                        },
                        Key::Up => self.ui.change_selected(MoveDestination::Up),
                        Key::Down => self.ui.change_selected(MoveDestination::Down),
                        Key::Left => self.ui.change_selected(MoveDestination::Left),
                        Key::Right => self.ui.change_selected(MoveDestination::Right),
                        _ => need_redraw = false
                    }
                }
                _ => need_redraw = false
            }
        } else {
            match button {
                Button::Keyboard(key) => {
                    match key {
                        Key::R => self.restart(),
                        Key::Up => self.field.move_selection(MoveDestination::Up),
                        Key::Down => self.field.move_selection(MoveDestination::Down),
                        Key::Left => self.field.move_selection(MoveDestination::Left),
                        Key::Right => self.field.move_selection(MoveDestination::Right),
                        Key::Space => {
                            let ind = self.field.get_selected_ind();
                            self.open_cell(ind);
                        },
                        Key::LCtrl | Key::RCtrl => {
                            let ind = self.field.get_selected_ind();
                            self.toggle_mark(ind);
                        }
                        Key::H => {
                            self.ui.proc_key(ParamType::Height);
                            self.in_ui = true;
                        },
                        Key::M =>{
                            self.ui.proc_key(ParamType::Mines);
                            self.in_ui = true;
                        },
                        Key::W => {
                            self.ui.proc_key(ParamType::Width);
                            self.in_ui = true;
                        },
                        _ => need_redraw = false
                    }
                },
                Button::Mouse(btn) => {
                    let field_rect = self.get_field_rect(window);
                    let cell_w = field_rect[2] / self.field.get_width();
                    let cell_h = field_rect[3] / self.field.get_height();
                    let mouse_x = self.mouse_x.floor() as u32;
                    let mouse_y = self.mouse_y.floor() as u32;
                    if (mouse_x < field_rect[0]) || (mouse_x > field_rect[0] + field_rect[2]) ||
                       (mouse_y < field_rect[1]) || (mouse_y > field_rect[1] + field_rect[3]) {
                        return;
                    }
                    let x = (mouse_x - field_rect[0]) / cell_w;
                    let y = (mouse_y - field_rect[1]) / cell_h;
                    let w = self.field.get_width();
                    match btn {
                        MouseButton::Left => {
                            self.open_cell(x + y*w);
                        },
                        MouseButton::Right => {

                            self.toggle_mark(x + y*w);
                        },
                        _ => need_redraw = false
                    }
                }
                Button::Joystick(_) => {
                    // Doen't work for me (Logitech Rumble F510)
                    // println!("{:?}", btn);
                }
            }
        }

        if need_redraw {
            self.render_state.reset();
        }
    }

    fn toggle_mark(&mut self, i: u32) {
        if self.game_ended || self.field.revealed(i) {
            return;
        }
        self.field.toggle_mark(i);
    }

    fn open_cell(&mut self, i: u32) {
        if self.game_ended {
            return;
        }
        if self.game_start == None {
            self.game_start = Some(UTC::now()); 
        }

        self.field.reset_if_need(i);

        if !self.field.revealed(i) {
            self.check_reveal(i);
        } else {
            self.reveal_neighborhood(i);
        }
    }

    fn reveal_neighborhood(&mut self, i: u32) {
        let cells_count_opt = {
            if let &Content::Number(cells_count) = self.field.get_content(i) {
                Some(cells_count)
            } else {
                None
            }
        };

        if let Some(cells_count) = cells_count_opt {
            if self.field.get_neighborhood_count(i) == cells_count {
                let on_left_edge = i % self.field.get_width() == 0;
                let on_right_edge = (i+1) % self.field.get_width() == 0; 

                for rdelta in -1..2i32 {
                    for cdelta in -1..2i32 {
                        if on_left_edge && (cdelta == -1) {
                            continue;
                        }
                        if on_right_edge && (cdelta == 1) {
                            continue;
                        }
                        let tgt = (i as i32) + rdelta*(self.field.get_width() as i32) + cdelta;
                        if (tgt < 0) || (tgt >= self.field.get_size() as i32) {
                            continue;
                        }
                        self.check_reveal(tgt as u32);
                    }
                }
            }
        }
    }

    pub fn check_reveal(&mut self, i : u32) {
        if self.field.marked(i) {
            return;
        }
        match *self.field.reveal(i) {
            Content::Mine(_) => {
                self.field.reveal_all();
                self.game_ended = true;
                self.field.set_killer(i);
                self.game_end = Some(UTC::now()); 
                self.msg.show( GameEndState::Lose );
                println!("Game over :(");
            },
            Content::None => {
                self.field.chain_reveal(i);
                if self.field.is_victory() {
                    self.msg.show( GameEndState::Win );
                    println!("You win :)");
                    self.game_ended = true;
                    self.game_end = Some(UTC::now()); 
                }
            }
            Content::Number(_i) => {
                if self.field.is_victory() {
                    self.msg.show( GameEndState::Win );
                    println!("You win :)");
                    self.game_ended = true;
                    self.game_end = Some(UTC::now()); 
                } 
            }
        }
    }

    pub fn mouse_move(&mut self, mouse_rel: [f64; 2]) {
        self.mouse_x = mouse_rel[0];
        self.mouse_y = mouse_rel[1];
    }

    pub fn update_state(&mut self) {
        if let Some(_) = self.game_start {
            if let None = self.game_end {
                let now = UTC::now();
                match self.last_time {
                    Some(last_time) => {
                        let time_from_last_tick = now - last_time;
                        if time_from_last_tick >= Duration::seconds(1) {
                            self.last_time = Some(now);
                            self.render_state.reset();
                        }
                    },
                    None => self.last_time = Some(now),
                }
            }
        }
    }

    pub fn resize(&mut self) {
        self.render_state.reset();
    }

    fn restart(&mut self) {
        self.game_ended = false;
        self.msg.hide();
        self.game_start = None;
        self.game_end = None; 
        self.last_time = None;
        self.field.restart();
        self.render_state.reset();
    }
}


src/main.rs
extern crate piston_window;
#[macro_use]
extern crate clap;
extern crate chrono;
extern crate rand;
extern crate find_folder;

mod game;
mod field;
mod ui;
mod common;

use piston_window::*;
use clap::{Arg, App};

fn main() {
    let matches = App::new("Minesweeper")
        .author("Alexander Kuvaev <alexander@kuvaev.me>")
        .version(&*format!("v{}", crate_version!()))
        .about("Simple minesweeper in Rust ")
        .arg(Arg::from_usage("-s, --size [size] 'window size \"width\"x\"height\"'")
            .validator(|val| {
                let mut s = val.split("x");
                match s.next() {
                    Some(w) => {
                        match w.parse::<u32>() {
                            Ok(r) => {
                                if (r < 100) || (r > 4096) {
                                    return Err(format!("window width must be between 100 and 4096"))
                                }
                            },
                            Err(e) => return Err(format!("window width arg error: {}", e))
                        }
                    }
                    None => return Err(format!("window size expected \"width\"x\"height\", got {}", val))
                }
                match s.next() {
                    Some(h) => {
                        match h.parse::<u32>() {
                            Ok(r) => {
                                if (r < 100) || (r > 3072) {
                                    return Err(format!("window height must be between 100 and 3072"))
                                }
                            },
                            Err(e) => return Err(format!("window height arg error: {}", e))
                        }
                    }
                    None => return Err(format!("window size expected \"width\"x\"height\", got {}", val))
                }
                match s.next() {
                    Some(_r) => Err(format!("window size arg expected \"width\"x\"height\", got {}", val)),
                    None => Ok(())
                }
            }))
        .arg(Arg::from_usage("-f, --field [field] 'field size \"width\"x\"height\"'")
            .validator(|val| {
                let mut s = val.split("x");
                match s.next() {
                    Some(w) => {
                        match w.parse::<u32>() {
                            Ok(r) => {
                                if (r < 5) || (r > 100) {
                                    return Err(format!("field width must be between 5 and 100"))
                                }
                            },
                            Err(e) => return Err(format!("field width arg error: {}", e))
                        }
                    }
                    None => return Err(format!("field size expected \"width\"x\"height\", got {}", val))
                }
                match s.next() {
                    Some(h) => {
                        match h.parse::<u32>() {
                            Ok(r) => {
                                if (r < 5) || (r > 100) {
                                    return Err(format!("field height must be between 5 and 100"))
                                }
                            },
                            Err(e) => return Err(format!("field height arg error: {}", e))
                        }
                    }
                    None => return Err(format!("field size expected \"width\"x\"height\", got {}", val))
                }
                match s.next() {
                    Some(_r) => Err(format!("field size arg expected \"width\"x\"height\", got {}", val)),
                    None => Ok(())
                }
            }))
        .arg(Arg::from_usage("-m, --mines [mines] 'max mines'"))
        .arg(Arg::from_usage("--oldOGL 'set OpenGL version to 2.1'"))
        .arg(Arg::from_usage("--maxFPS [max_fps] 'set max fps'"))
        .get_matches();

    let mut width = 1024;
    let mut height = 768;
    let mut f_width = 20;
    let mut f_height = 20;
    let mut mines = 50;
    let mut max_fps = 60;
    let mut opengl = OpenGL::V3_2;
    if let Some(size) = matches.value_of("size") {
        let mut s = size.split("x");
        width = s.next().unwrap().parse().unwrap_or(width);
        height = s.next().unwrap().parse().unwrap_or(height);
    }
    if let Some(field) = matches.value_of("field") {
        let mut s = field.split("x");
        f_width = s.next().unwrap().parse().unwrap_or(f_width);
        f_height = s.next().unwrap().parse().unwrap_or(f_height);
    }
    if let Some(m) = matches.value_of("mines") {
        mines = m.parse().unwrap_or(mines);
    }
    if let Some(m) = matches.value_of("max_fps") {
        max_fps = m.parse().unwrap_or(max_fps);
    }
    if matches.is_present("oldOGL") {
        opengl = OpenGL::V2_1;
    }

    let mut window: PistonWindow =
        WindowSettings::new("Minesweeper", [width, height])
        .opengl(opengl)
        .exit_on_esc(true)
        .build()
        .unwrap();

    let assets = find_folder::Search::ParentsThenKids(3, 3)
        .for_folder("assets").unwrap();
    let ref font = assets.join("FiraSans-Regular.ttf");
    let factory = window.factory.borrow().clone();
    let glyphs = Glyphs::new(font, factory).unwrap();

    let mut game = game::Game::new(glyphs, f_width, f_height, mines);
    window.set_max_fps(max_fps);
    for e in window {
        game.update_state();

        if let Some(_) = e.render_args() {
        	game.render(&e);
        }

        if let Some(_) = e.resize_args() {
        	game.resize()
        }

        if let Some(mouse_rel) = e.mouse_cursor_args() {
            game.mouse_move(mouse_rel);
        }

        if let Some(button) = e.press_args() {
            game.proc_key(button, &e);
        }

    }
}
